Sources/ServiceLifecycle/Docs.docc/How to adopt ServiceLifecycle in libraries.md

```diff
@@ -0,0 +1,126 @@
+the `Task`. At this point, we have just reinvented Structured Concurrency.
+To avoid all of these problems we can just conform to the ``Service`` protocol which requires a `run()`
+method. This requirements already guides us to implement the long running work inside the `run()` method.
+Having this method allows the user of the client to decide in which task to schedule the keep-alive pings.
+They can still decide to create an unstructured `Task` for this, but that is up to the user now. 
+Furthermore, we now get automatic cancellation propagation from the task that called our `run()` method.
+Below is an overhauled implementation that exposes such a `run()` method.
+
+```swift
+public actor TCPEchoClient: Service {
+  public init() { }
+
+  public func run() async throws {
+    while true {
+      self.sendKeepAlivePings()
+      try await Task.sleep(for: .second(1))
+    }
+  }
+
+  private func sendKeepAlivePings() async { ... }
+}
+```
+
+
+### Returning from your `run()` method
+
+Since the `run()` method contains long running, work returning from it is seen as a failure and will
+lead to the ``ServiceRunner`` cancelling all other services by cancelling the task that is running
+their respective `run()` method.
+
+### Cancellation
+
+Structured Concurrency propagates task cancellation down the task tree. Every task in the tree can
+check for cancellation or react to it with cancellation handlers. The ``ServiceRunner`` is using task
+cancellation to tear everything down in the case of an early return or thrown error from the `run()`
+method of any of the services. Hence it is important that each service properly implements task
+cancellation in their `run()` methods.
+
+Note: If your `run()` method is only calling other async methods that support cancellation themselves
+or is consuming an `AsyncSequence`, you don't have to do anything explicitly here. Looking at the
+`TCPEchoClient` example from above we can see that we only call `Task.sleep` in our `run()` method
+which is supporting task cancellation.
+
+### Graceful shutdown
+
+When running an application in a real environment it is often required to gracefully shutdown the application.
+For example the application might be running in Kubernetes and a new version of it got deployed. In this
+case, Kubernetes is going to send a `SIGTERM` signal to the application and expects it to terminate
+within a grace period. If the application isn't stopping in time then Kubernetes will send the `SIGKILL`
+signal and forcefully terminate the process.
+For this reason ``ServiceLifecycle`` introduce a new _shutdown gracefully_ concept that allows to
+terminate the work in a structured and graceful manner. This works similar to task cancellation but
+it is fully opt in and up to the business logic of the application to decide what to do.
+
+``ServiceLifecycle`` exposes one free function called ``withShutdownGracefulHandler(operation:onGracefulShutdown:)``
```

Comment:
It would be useful to have an example here of how a service would incorporate this method.